### PR TITLE
[DeviceSanitizer] Force split per kernel if device sanitizer is enabled

### PR DESCRIPTION
**Pull request by zhaomaosu**
Due to the limitations of device sanitizer implementation, we need to always split device code per kernel if device sanitizer is enabled.

### CODE REVIEW

**Review by AllanZyne — APPROVED** [review]

File: clang/test/Driver/sycl-device-sanitizer.cpp [hunk 1]
```diff
@@ -15,3 +15,17 @@
...
+// RUN: %clangxx -fsycl -Xarch_device -fsanitize=address %s -### 2>&1 \
+// RUN:   | FileCheck --check-prefix=SYCL-ASAN-SPLIT %s
+
+// RUN: %clangxx -fsycl -Xsycl-target-frontend -fsanitize=address %s -### 2>&1 \
+// RUN:   | FileCheck --check-prefix=SYCL-ASAN-SPLIT %s
+
+// RUN: %clangxx -fsycl -fsanitize=address -fsycl-device-code-split=per_source %s -### 2>&1 \
+// RUN:   | FileCheck --check-prefix=SYCL-ASAN-SPLIT %s
+
+// SYCL-ASAN-SPLIT: sycl-post-link{{.*}} "-split=kernel"
```
Comment (wenju-he) [inline clang/test/Driver/sycl-device-sanitizer.cpp:31]: no end-of-line here

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -10635,8 +10635,50 @@ static void getNonTripleBasedSYCLPostLinkOpts(const ToolChain &TC,
...
+      // -Xsycl-target-frontend
+      if (llvm::any_of(TCArgs.getAllArgValues(options::OPT_Xsycl_frontend),
+                       [](const std::string &Arg) {
+                         return Arg.find("-fsanitize=address") !=
+                                std::string::npos;
+                       }))
+        return true;
+
+      // -Xsycl-target-frontend=
+      if (llvm::any_of(TCArgs.getAllArgValues(options::OPT_Xsycl_frontend_EQ),
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:10659]: There is no scrutiny on any specific target as specified with `-Xsycl-target-frontend=<target> <opt>`.  This case is probably not used much (if at all) but we should at least call out the discrepancy in a comment.

File: clang/test/Driver/sycl-device-sanitizer.cpp [hunk 3]
```diff
@@ -15,3 +15,17 @@
...
 // SYCL-ASAN-FILTER-SAME: "-mllvm" "-asan-stack=0"
+
+// RUN: %clangxx -fsycl -fsanitize=address %s -### 2>&1 \
+// RUN:   | FileCheck --check-prefix=SYCL-ASAN-SPLIT %s
+
+// RUN: %clangxx -fsycl -Xarch_device -fsanitize=address %s -### 2>&1 \
+// RUN:   | FileCheck --check-prefix=SYCL-ASAN-SPLIT %s
+
+// RUN: %clangxx -fsycl -Xsycl-target-frontend -fsanitize=address %s -### 2>&1 \
+// RUN:   | FileCheck --check-prefix=SYCL-ASAN-SPLIT %s
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-device-sanitizer.cpp:26]: Please also add a test for `-Xsycl-target-frontend=<target> -fsanitize=address`

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (zhaomaosu) [reply clang/lib/Driver/ToolChains/Clang.cpp:10659]: Updated the comments

File: clang/test/Driver/sycl-device-sanitizer.cpp (hunk 3 above)
Comment (zhaomaosu) [reply clang/test/Driver/sycl-device-sanitizer.cpp:26]: Done

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:10659]: Apologies - it looks like my original message didn't provide enough detail of what I was looking for.  The issue is that the enabling of the sanitizer is only gated by using `-Xsycl-target-frontend=<target> -fsanitize=address`.  This means no matter what `<target>` value is used, `-fsanitize-address` will be passed to all device compilations.

For example, `-fsycl -fsycl-targets=spir64_gen,spir64_x86_64,spir64 -Xsycl-target-frontend=spir64 -fsanitize=address` will apply to `spir64_gen` and `spir64_x86_64` as well, even if that is not the users intention.

Comment (zhaomaosu) [reply clang/lib/Driver/ToolChains/Clang.cpp:10659]: @mdtoguchi, thanks for your patiently explaining, it's clear to me. But we have found a more reasonable fix for the original problem. So, we don't need this PR anymore. Anyway, thanks for your comments again.

**Pull request closed without merging**